server/Arcadia.Assistant.CSP/Vacations/VacationsSyncExecutor.cs

```diff
@@ -0,0 +1,185 @@
+﻿namespace Arcadia.Assistant.CSP.Vacations
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    using Microsoft.EntityFrameworkCore;
+
+    using Arcadia.Assistant.Calendar.Abstractions;
+    using Arcadia.Assistant.CSP.Model;
+
+    public class VacationsSyncExecutor
+    {
+        private readonly Func<ArcadiaCspContext> contextFactory;
+
+        public VacationsSyncExecutor(Func<ArcadiaCspContext> contextFactory)
+        {
+            this.contextFactory = contextFactory;
+        }
+
+        public async Task<IEnumerable<Vacation>> UpsertVacation(CalendarEvent @event, VacationApprovals approvals, VacationsMatchInterval matchInterval)
+        {
+            var vacation = this.CreateVacationFromCalendarEvent(@event, approvals);
+
+            using (var context = this.contextFactory())
+            {
+                var existingVacations = await this.GetVacations(
+                    context,
+                    int.Parse(@event.EmployeeId),
+                    matchInterval.Start,
+                    matchInterval.End);
+
+                if (existingVacations.Count == 0)
+                {
+                    var createdVacation = (await context.Vacations.AddAsync(vacation)).Entity;
+                    existingVacations = new List<Vacation> { createdVacation };
```

Comment:
not really important, but I'm not happy with creating a new list instead of .Add(createdVacation). I think semantically it's cleaner to not change existing reference

Reply:
Sure you're right, especially because it was `HashSet` by default, and becomes `List` after this line.